juju_verify/verifiers/ceph.py

```diff
@@ -330,13 +528,65 @@ def _get_ceph_mon_app_map(self) -> Dict[str, Unit]:
         """
         applications = {unit.application for unit in self.units}
         logger.debug("affected applications %s", ", ".join(applications))
+        return {name: self._get_ceph_mon_unit(name) for name in applications}
+
+    def _find_units_in_ceph_tree(
+        self, tree: CephTree, device_class: Optional[str] = None
+    ) -> Set[Unit]:
+        """Get set of units from ceph tree."""
+        units = set()
+        for node in tree.nodes:
+            if node.type == "host" and node.children is not None:
+                if device_class is not None:
+                    # check if there is any osd with required device_class
+                    if not any(
+                        osd.device_class == device_class
+                        for osd in tree.nodes
+                        if osd.id in node.children
+                    ):
+                        continue
+
+                units.add(find_unit_by_hostname(self.model, node.name, self.NAME))
+
+        return units
 
-        app_map = {name: self._get_ceph_mon_unit(name) for name in applications}
-        logger.debug(
-            "found units %s", ", ".join([str(unit) for unit in app_map.values()])
-        )
+    @staticmethod
+    def _count_branch(tree: CephTree, units: Set[Unit], root_type: str) -> int:
+        """Count unique branches for list of units (hots) in Ceph Tree."""
+        if root_type == "host":
+            return len(units)
+
+        ancestors = set()
+        for unit in units:
+            node = tree.get_node(unit.machine.hostname)
+            ancestor = tree.find_ancestor(node, root_type)
+            if ancestor is None:
+                raise CharmException(f"Could nod find ancestors for node `{node.name}`")
+
+            logging.debug("Found ancestor %s for unit %s", ancestor, unit)
+            ancestors.add(ancestor)
+
+        return len(ancestors)
+
+    def check_ceph_pools(self) -> Result:
+        """Check whether Ceph cluster pools meet the requirements."""
+        for ceph_mon_unit in self.ceph_mon_app_map.values():
+            pools = self.get_ceph_pools(ceph_mon_unit)
+
+            if any(pool.type != 1 for pool in pools):
```

Comment:
Please add comments. I imagine type==1 is replicated, while type==0 is erasure-coded? The message returned seems to imply this, but let's make it explicit.

Reply:
There was a comment, but I remove it because the message and complex docstring. But I'll add it back.